blakserv/ccode.c

```diff
@@ -2344,3 +2350,199 @@ blak_int C_MinigameStringToNumber(int object_id,local_var_type *local_vars,
+   col_val = RetrieveValue(object_id, local_vars, normal_parm_array[2].type, normal_parm_array[2].value);
+   fr_val = RetrieveValue(object_id, local_vars, normal_parm_array[3].type, normal_parm_array[3].value);
+   fc_val = RetrieveValue(object_id, local_vars, normal_parm_array[4].type, normal_parm_array[4].value);
+   id_list_val = RetrieveValue(object_id, local_vars, normal_parm_array[5].type, normal_parm_array[5].value);
+
+   if (room_val.v.tag != TAG_ROOM_DATA)
+   {
+      bprintf("C_PointInSector has bad room tag\n");
+      return NIL;
+   }
+
+   if (row_val.v.tag != TAG_INT)
+   {
+      bprintf("C_PointInSector has bad row tag\n");
+      return NIL;
+   }
+
+   if (col_val.v.tag != TAG_INT)
+   {
+      bprintf("C_PointInSector has bad col tag\n");
+      return NIL;
+   }
+
+   if (fr_val.v.tag != TAG_INT)
+   {
+      bprintf("C_PointInSector has bad fr tag\n");
+      return NIL;
+   }
+
+   if (id_list_val.v.tag == TAG_INT)
+   {
+      val_type temp, nil_val;
+      temp.v.tag = TAG_INT;
+      temp.v.data = id_list_val.v.data;
+
+      nil_val.v.tag = TAG_INT;
+      nil_val.v.data = 0;
+
+      id_list_val.v.data = Cons(temp, nil_val);
+      id_list_val.v.tag = TAG_LIST;
+   }
+   else if (id_list_val.v.tag != TAG_LIST)
+   {
+      bprintf("C_PointInSector has bad 6th parameter tag, expected list or int\n");
+      return NIL;
+   }
+
+   roomdata_node *rd = GetRoomDataByID(room_val.v.data);
+   if (!rd || !rd->file_info.sectors)
+   {
+      bprintf("C_PointInSector has bad room id passed in: %i\n", room_val.v.data);
+      return NIL;
+   }
+
+   room_type *r = &rd->file_info;
+
+   // remember that kod uses 1-based arrays and we don't
+   int row0 = (int) row_val.v.data - 1;
+   int col0 = (int) col_val.v.data - 1;
+
+   // Now, we need to convert the kod coordinates to coordinates that match the room data
+   int fine_x = (int) ((long long) (col0 * FINENESS) + fc_val.v.data) * POLYGON_COORD_SCALE;
+   int fine_y = (int) ((long long) (row0 * FINENESS) + fr_val.v.data) * POLYGON_COORD_SCALE;
+
+   // Build the vector of sector ids
+   if (id_list_val.v.tag == TAG_INT)
+   {
+      sector_id = (int) id_list_val.v.data;
+      sector_ids.push_back(sector_id);
+   }
+   else if (id_list_val.v.tag == TAG_LIST)
+   {
+      list_node *list = GetListNodeByID(id_list_val.v.data);
+      if (!list)
+      {
+         bprintf("C_PointInSector failed to get list node by id: %i\n", id_list_val.v.data);
+         return NIL;
+      }
+
+      while (list != NULL)
+      {
+         if (list->first.v.tag == TAG_INT)
+         {
+            sector_id = list->first.v.data;
+            sector_ids.push_back(sector_id);
+         }
+         else
+         {
+            bprintf("C_PointInSector: Sector list element is not an integer\n");
+         }
+
+         if (list->rest.v.tag == TAG_LIST)
+         {
+            list = GetListNodeByID(list->rest.v.data);
+         }
+         else
+         {
+            break;

```

Comment:
This is an error, no?

Reply:
My intent was to handle the end of the list by simply breaking and moving on, using the sector ids already collected. But yes, it _could_ be an error, or it could simply be the end of the list.